### ISSUE TITLE
Analyzer produces StackOverflow error in some cases when typedef references itself directly.

### ISSUE THREAD

**Issue by iarkh** [dart-lang/sdk#33599]
* Dart SDK Version: 2.0.0-dev.64.1
* OS: Windows10

Dartanalyzer produces unexpected StackOverflow error in some difficult cases when typedef references itself directly.

Here is a simple code example:
```
typedef F<X extends F> = F Function();
main() {}
```
Sample output is:
```
$>dart --version
Dart VM version: 2.0.0-dev.64.1 (Thu Jun 21 08:47:55 2018 +0200) on "windows_x64"

$>dartanalyzer --preview-dart-2 test.dart
Analyzing test.dart...
Unhandled exception:
Stack Overflow
#0      DynamicTypeImpl.substitute2 (package:analyzer/src/dart/element/type.dart:343:12)
#1      TypeImpl.substitute (package:analyzer/src/dart/element/type.dart:2575:12)
#2      _FunctionTypeImplLazy.substitute2 (package:analyzer/src/dart/element/type.dart:3184:18)
#3      _FunctionTypeImplLazy.returnType (package:analyzer/src/dart/element/type.dart:3047:41)
#4      FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:657:34)
#5      FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:709:34)
#6      FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:709:34)
#7      FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:709:34)
#8      FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:709:34)
<very long stack trace here>
#6091   FunctionTypeImpl.appendTo (package:analyzer/src/dart/element/type.dart:709:34)
#6092   TypeImpl.toString (package:analyzer/src/dart/element/type.dart:2503:5)
#6093   _StringBase._interpolateSingle (dart:core/runtime/libstring_patch.dart:799:17)
#6094   StringBuffer.write (dart:core/runtime/libstring_buffer_patch.dart:64:24)
#6095   TypeParameterElementImpl.appendTo (package:analyzer/src/dart/element/element.dart:9633:14)
#6096   GenericTypeAliasElementImpl.appendTo (package:analyzer/src/dart/element/element.dart:5625:57)
#6097   ElementImpl.toString (package:analyzer/src/dart/element/element.dart:3607:5)
#6098   _StringBase._interpolate (dart:core/runtime/libstring_patch.dart:817:19)
#6099   new _ElementMismatchException (package:analyzer/src/generated/declaration_resolver.dart:1016:65)
#6100   DeclarationResolver.resolve (package:analyzer/src/generated/declaration_resolver.dart:74:17)
#6101   LibraryAnalyzer._resolveFile (package:analyzer/src/dart/analysis/library_analyzer.dart:688:10)
#6102   LibraryAnalyzer._analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:130:9)
#6103   __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.forEach (dart:collection/runtime/libcompact_hash.dart)
#6104   LibraryAnalyzer._analyze (package:analyzer/src/dart/analysis/library_analyzer.dart:129:13)
#6105   LibraryAnalyzer.analyze.<anonymous closure> (package:analyzer/src/dart/analysis/library_analyzer.dart:104:16)
#6106   _RootZone.runUnary (dart:async/zone.dart:1381:54)
#6107   _FutureListener.handleValue (dart:async/future_impl.dart:129:18)
#6108   Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:638:45)
#6109   Future._propagateToListeners (dart:async/future_impl.dart:667:32)
#6110   Future._addListener.<anonymous closure> (dart:async/future_impl.dart:347:9)
#6111   _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#6112   _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#6113   _runPendingImmediateCallback (dart:isolate/runtime/libisolate_patch.dart:113:13)
#6114   _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:166:5)

#0      AnalysisDriver._computeAnalysisResult.<anonymous closure> (package:analyzer/src/dart/analysis/driver.dart:1323:9)
<asynchronous suspension>
#1      PerformanceLog.runAsync (package:front_end/src/base/performance_logger.dart:52:21)
<asynchronous suspension>
#2      AnalysisDriver._computeAnalysisResult (package:analyzer/src/dart/analysis/driver.dart:1240:20)
<asynchronous suspension>
#3      AnalysisDriver.getErrors (package:analyzer/src/dart/analysis/driver.dart:652:43)
<asynchronous suspension>
#4      AnalyzerImpl.prepareErrors (package:analyzer_cli/src/analyzer_impl.dart:130:58)
<asynchronous suspension>
#5      AnalyzerImpl._analyze (package:analyzer_cli/src/analyzer_impl.dart:170:11)
<asynchronous suspension>
#6      AnalyzerImpl.analyze (package:analyzer_cli/src/analyzer_impl.dart:106:18)
<asynchronous suspension>
#7      Driver._runAnalyzer (package:analyzer_cli/src/driver.dart:763:21)
#8      Driver._analyzeAllImpl (package:analyzer_cli/src/driver.dart:384:40)
#9      _RootZone.runUnary (dart:async/zone.dart:1381:54)
#10     _FutureListener.handleValue (dart:async/future_impl.dart:129:18)
#11     Future._propagateToListeners.handleValueCallback (dart:async/future_impl.dart:638:45)
#12     Future._propagateToListeners (dart:async/future_impl.dart:667:32)
#13     Future._addListener.<anonymous closure> (dart:async/future_impl.dart:347:9)
#14     _microtaskLoop (dart:async/schedule_microtask.dart:41:21)
#15     _startMicrotaskLoop (dart:async/schedule_microtask.dart:50:5)
#16     _runPendingImmediateCallback (dart:isolate/runtime/libisolate_patch.dart:113:13)
#17     _RawReceivePortImpl._handleMessage (dart:isolate/runtime/libisolate_patch.dart:166:5)
```
Please note that:

1 . Dart crashes on the same example with the same stack trace as it is described in the bug #33152. Is it the same issue?
2. `typedef F = F Function();` fails with compiler error as expected
3. `typedef F<X extendss F> = void Function();` crashes with another stack trace like it's described in the bug #33152.

**Comment by iarkh** [comment]
Cannot reproduce with Dev.64.0

**Comment by eernstg** [comment]
Apart from the fact that no tools should loop or crash, it's worth considering the correct behavior.

Following the treatment of other declarations, the name of the type alias, `F`, is added to the library scope (that is, the enclosing scope for the type alias declaration itself); then we use that `F` as a raw bound for `X`. This should be a compile-time error because `F` does not have simple bounds.

So the VM should just reject the program, as should the analyzer.

**Comment by MichaelRFairhurst** [comment]
https://dart-review.googlesource.com/c/sdk/+/76040 fixes this, but interacts perhaps poorly with the fix to #33740.
